### ISSUE TITLE
sql: Timestamp ops work with ns prec although type says µs prec - causes invalid truncation and errors

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#8755]
See #8590 for some discussion.

Go's `time.Duration` stores duration information at a nanosecond resolution, meaning that any computations which pass through the type are bounded to `±290 years`. This affects operations like `timestamp - timestamp`, which will now overflow if they reach the `time.Duration` limit. Postgres performs all time-related operations at the microsecond resolution, meaning that it's duration limits are 1000x greater. In an effort to maintain consistency with Postgres, we should investigate if it is worth the effort to increase this limit, either by performing the operation and returning a result in the microsecond resolution or by finding another alternative.

@dt @paperstreet @mjibson I believe you were all involved in the previous discussion of parsing timestamps with or without nanoseconds included. If you have any extra context for this issue, feel free to add it below.

**Comment by mjibson** [comment]
What happens now in the case the math is greater than the duration can handle? Undetected overflow? We recently fixed a few other cases of overflow by detecting it and throwing an error. If we are not doing that, we should at least detect and error on overflow.

I think supporting larger durations than 290 years is not a super useful feature. I think detect + throw error for now. If users complain and have valid use cases then we can come back and reconsider. I say that because I think it will be difficult to correctly support anything else while maintaining nanosecond precision, which I don't want to lose.

**Comment by nvanbenschoten** [comment]
I think the correct term for it would be "saturating". If the operation would overflow a `time.Duration`, it just returns the min/max value instead. This makes it easy to catch and return an error instead. Like you mentioned, that's probably the best course of action for now.

**Comment by paperstreet** [comment]
=mjibson

**Comment by mjibson** [comment]
Although saturation happens in some cases, it is easily thwarted.

```
> select interval '1h' * 24 * 365 * 300;
+--------------------------------+
| ((CAST('1h' AS INTERVAL) * 24) * 365) * 300 |
+--------------------------------+
| -2496095h34m33.709551616s      |
+--------------------------------+
root@:26257> select timestamp '2000-01-01' - timestamp '1700-01-01';
+--------------------------------+
| CAST('2000-01-01' AS TIMESTAMP) - CAST('1700-01-01' AS TIMESTAMP) |
+--------------------------------+
| 2562047h47m16.854775807s       |
+--------------------------------+
(1 row)
root@:26257> select timestamp '2000-01-01' - timestamp '1700-01-01' + interval '1s';
+--------------------------------+
| (CAST('2000-01-01' AS TIMESTAMP) - CAST('1700-01-01' AS TIMESTAMP)) + CAST('1s' AS INTERVAL) |
+--------------------------------+
| -2562047h47m15.854775809s      |
+--------------------------------+
```

We need to add overflow/saturation detection in more places then.

**Comment by dt** [comment]
my guess is that overwhelming majority of users do not need sub-micro precision, but they probably also don't need 300 year durations either. that said, we also have hacks scattered around to handle the fact that timestamps on in postgres and thus most clients express only micros, and thus to be round-trip safe, we have to truncate anyway. my instinct would be to default to micros everywhere to eliminate strange surprises and gotchas, and add the nano-speicifc helpers where anyone actually wants to use nanos.

**Comment by dt** [comment]
as far as losing "maintaining nanosecond precision", since we already default to truncating nanos for pg wire round-trips, it seems like keeping the operators in nanos is just moving a bunch of zeros back and forth at the expense of usable capacity at the other end.

**Comment by vivekmenezes** [comment]
@mjibson this is fixed right?

**Comment by mjibson** [comment]
I don't think this is fixed: the queries I posted above still return incorrect results.

**Comment by knz** [comment]
confirmed this still repros in 2.1-alpha. 

The problem is not related to the interaction between timestamps and intervals though; the error detected by matt initially can be readily reproduced with:

```
> select interval '2562047h47m16s854ms775us807ns' + interval '1s';
+-------------------------------------+
|   '2562047h47m16s854ms775µs807ns'   |
|               + '1s'                |
+-------------------------------------+
| -2562047h-47m-15s-854ms-775µs-809ns |
+-------------------------------------+
```

The problem is that the arithmetic on durations first invokes `duration.normalize()`, and that causes the overflow. We should be able to do arithmetic on durations (intervals) without normalization.

**Comment by asubiotto** [comment]
I believe this has been fixed by #43292
